### ISSUE TITLE
addic7ed provider broken on 1.x for titles containing characters other than letters and digits

### ISSUE THREAD

**Issue by miles82** [Diaoul/subliminal#500]
I recently upgraded to 1.x , and now I'm getting an error:

```
DEBUG:subliminal.providers.addic7ed:Getting show id
INFO:subliminal.providers.addic7ed:Getting the page of show id 4863, season 1
ERROR:subliminal.api:Unexpected error in provider 'addic7ed', discarding it
Traceback (most recent call last):
  File "/volume2/homes/admin/srt_venv/lib/python2.7/site-packages/subliminal/api.py", line 117, in list_subtitles
    provider_subtitles = self[name].list_subtitles(video, provider_languages)
  File "/volume2/homes/admin/srt_venv/lib/python2.7/site-packages/subliminal/providers/addic7ed.py", line 257, in list_subtitles
    return [s for s in self.query(video.series, video.season, video.year)
  File "/volume2/homes/admin/srt_venv/lib/python2.7/site-packages/subliminal/providers/addic7ed.py", line 227, in query
    series = match.group('series')
AttributeError: 'NoneType' object has no attribute 'group'
```

It's because of this regex in [addic7ed.py](https://github.com/Diaoul/subliminal/blob/master/subliminal/providers/addic7ed.py#L18).

    series_year_re = re.compile('^(?P<series>[ \w]+)(?: \((?P<year>\d{4})\))?$')

It allows only spaces, letters and digits in show title. The one I'm looking for also has an apostrophe, but there are many other characters that show up in titles regularly like dots, colons, dashes, etc. It would be better if only the year part was detected with regex and everything before it treated as show title (so you don't have to enumerate all possible characters).

**Comment by Diaoul** [comment]
What is the show?

**Comment by miles82** [comment]
This was [Survivor's Remorse](http://www.addic7ed.com/show/4863). Also [Mr. Robot](http://www.addic7ed.com/show/5151), etc.

**Comment by miles82** [comment]
I doublechecked and Mr. Robot actually fails even earlier with this error:

```
DEBUG:subliminal.providers.addic7ed:Getting show id
WARNING:subliminal.providers.addic7ed:Series not found in show ids
ERROR:subliminal.providers.addic7ed:No show id found for u'Mr Robot' ({u'country': None, u'year': None})
```

**Comment by h3llrais3r** [comment]
I had the same problem and for me it was related to the fact the episode was in a folder which also contains the series info.
Try to put it in a folder without the series info and it should work (at least it worked for me).
If it works, it's probably related to a bug in guessit (https://github.com/wackou/guessit/issues/190), which will be fixed in the next release 0.11.

**Comment by miles82** [comment]
@h3llrais3r I don't think it's related to guessit.

It looks to be down to [this line](https://github.com/Diaoul/subliminal/blob/master/subliminal/providers/addic7ed.py#L164).

I added:
```
logger.debug("SUGGESTION: " + suggestion[0].i.text.lower())
logger.debug("SERIES: " + series_year.lower()) 
```

And I get:
```
INFO:subliminal.providers.addic7ed:Searching show ids with {u'search': u'Mr Robot', u'Submit': u'Search'}
DEBUG:subliminal.providers.addic7ed:SUGGESTION: mr. robot
DEBUG:subliminal.providers.addic7ed:SERIES: mr robot
WARNING:subliminal.providers.addic7ed:Show id not found: suggestion does not match
ERROR:subliminal.providers.addic7ed:No show id found for u'Mr Robot' ({u'country': None, u'year': None})
```

The addic7ed suggestion is "Mr. Robot", and the series_year string is "Mr Robot" (no dot).

I see apostrophe's are cleaned [here](https://github.com/Diaoul/subliminal/blob/master/subliminal/providers/addic7ed.py#L133), perhaps the same should be done with at least the dot too. That way we find the match before performing search on addic7ed.

So to recap, the quick and dirty way to download "Mr. Robot" subtitles from addic7ed:
1. change `series_year_re` regex to allow dots
2. add `.replace('.', '')` in `_get_show_ids()`

**Comment by Diaoul** [comment]
Can you provide the video names so I can add unittests? I believe guessit will return `Mr Robot` and not `Mr. Robot`, same goes for `Survivors Remorse` instead of `Survivor's Remorse`.

Have you tested your solution?

**Comment by miles82** [comment]
The filenames are standard:
Mr.Robot.S01E01....mp4
Survivors.Remorse.S01E01...mp4

They don't include extra dots or apostrophes so guessit doesn't return them.

I tested my quick fix on my videos and it works, but I haven't run the unit tests. But as I said earlier the proper fix should probably not use a regex for the show's title and the replace in _get_show_ids() should remove other charachters (dashes, colons, parenthesis?) which are unlikely to be returned from guessit. Also the comparison in _search_show_id() would ideally be more lenient.

**Comment by Diaoul** [comment]
The comparison with search show id is made when computing subtitle score
and doesn't belong here. I don't want to include characters in the regex or
the replace if there's no unit test for it.

How would you extract show titles and year from the page without a regex?

**Comment by miles82** [comment]
Show title on addic7ed has the format "{title} ({year}) subtitles". The year part is optional. I'd:
1. remove " subtitles" from the end (you do that with [:-10]), then
2. check if remainder endswith (\d{4}), if yes save that as year and remove it from the string
3. the remainder is the show title

Are there any docs on how to run subliminal's tests?

**Comment by Diaoul** [comment]
If you do that, the show title would be correct (Mr. Robot) but different
than the video show name guessed from guessit (Mr Robot) and subtitle and
video won't match.

To fix that you can use a dedicated function that compares show names
removing special characters instead of standard equality check. I'm not
sure this is the best approach since it adds complexity.

**Comment by miles82** [comment]
Titles on addic7ed will contain dots, apostrophe's, dashes - we can't change that.
File names don't contain those characters, so guessit won't return them - we can't change that either.

Some complexity must be added and it's definitely worth it. From 3500+ shows on addic7ed almost a 1000 won't match with the current implementation.

**Comment by Diaoul** [comment]
I added some logic to subliminal on my local branch and ended up with good results. I'll push that in a week or so when I can get Addic7ed to work without getting 304s.

**Comment by Diaoul** [comment]
Fixed with bc218fa13848182c167817eeba5bd81df719a7ac
